New Control - ViewPicker - Fixing Issue 1439

NishkalankBezawada opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | [ ]
| New feature?    | [x]
| New sample?      | [ ]
| Related issues?  | fixes #1439 

#### What's in this Pull Request?

Added a new control 'ViewPicker' which enables to select the list views just like ListPicker, SitePicker, FieldPicker controls

#### Properties used

| Property | Type | Required | Description |
| ---- | ---- | ---- | ---- |
| context | BaseComponentContext | yes | The context object of the SPFx loaded webpart or customizer. |
| className | string | no | If provided, additional class name to provide on the dropdown element. |
| disabled | boolean | no | Whether or not the view picker control is disabled. |
| filter | string | no | Filter views from Odata query. |
| label | string | no | Label to use for the control. |
| listId | string | no | The List Id of the list. |
| placeholder | string | no | Placeholder label to show in the dropdown. |
| orderBy | Enum | no | How to order the set of views (By ID or Title). |
| selectedView | string OR string[] | no | Keys(View Ids) of the selected item(s). If you provide this, you must maintain selection state by observing onSelectionChanged events and passing a new value in when changed. |
| multiSelect | boolean | no | Optional mode indicates if multi-choice selections is allowed. Default to `false`. |
| showBlankOption | boolean | no | Whether or not to show a blank option. Default to `false`. |
| viewsToExclude | string[] | no | Defines view titles which should be excluded from the view picker control. |
| webAbsoulteUrl | string | no | Absolute Web Url of target site (user requires permissions) |
| onSelectionChanged | (newValue: string OR string[]): void | no | Callback function when the selected option changes. |



#### Screens

Below is the implementation of the control

```TypeScript
    <ViewPicker context={this.props.context}
        listId={"9f3908cd-1e88-4ab3-ac42-08efbbd64ec9"}
        placeholder={'Select list view(s)'}
        orderBy={orderBy.Title}
        multiSelect={true}
        onSelectionChanged={this.onViewPickerChange} />
```
![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/4623af73-615b-4db1-9ac3-9c4dcb88c2d2)


- The `onSelectionChanged` change event returns the selected view(s) and can be implemented as follows in your webpart:

```TypeScript
  private onViewPickerChange = (views: string | string[]) => {
    console.log("Views:", views);
  }
```


Below is the screen of the control from workbench

![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/6c36758b-a5b9-4317-8417-ea44c0245b22)


Below is the screen of the control in single selection mode

![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/5c99dd83-b07c-4400-95da-de401d1ea0bd)


Below is the screen of the control in multi selection mode

![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/f9c50896-2e6b-4854-a66a-ac24b895ad27)


Below are the sample screen of the list views. 

![image](https://user-images.githubusercontent.com/47456098/230456628-982d1b8f-d896-4c2e-aaf7-4a0e37eda165.png)


//Nishkalank Bezawada

NishkalankBezawada commented on package-lock.json:
Hello,

package-lock.json is not required to be merged in this commit. Kindly resolve this conflicts. 

//Nishkalank

NishkalankBezawada commented:
Hello Alex @AJIXuMuK

I was a bit away with some medical stuff. I will soon update the requested changes. 

Thanks & Have a wonderful weekend,
Nishkalank

NishkalankBezawada reviewed (commented):
Corrected as suggested. Kindly review. 

Thanks,
Nishkalank

NishkalankBezawada reviewed (commented):
Hello @AJIXuMuK 

Corrected and completed the changes as suggested. Kindly review.

Thanks,
Nishkalank

NishkalankBezawada commented:
Hello @AJIXuMuK @joelfmrodrigues 

Corrected and completed the changes as suggested. Kindly review.

Thanks,
Nishkalank

AJIXuMuK commented on src/common/utilities/GeneralHelper.ts:
```diff
@@ -4,6 +4,7 @@ import * as strings from 'ControlStrings';
 
 export const IMG_SUPPORTED_EXTENSIONS = ".gif,.jpg,.jpeg,.bmp,.dib,.tif,.tiff,.ico,.png,.jxr,.svg";
 
+import * as _ from '@microsoft/sp-lodash-subset';
```
Why do you need this `import`

NishkalankBezawada replied:
Hello @AJIXuMuK,

Removed this unwanted ```import``` in the latest commit

Thanks,
Nishkalank

AJIXuMuK commented on src/controls/viewPicker/IViewPicker.ts:
```diff
@@ -0,0 +1,100 @@
+import { BaseComponentContext } from '@microsoft/sp-component-base';
+import { ISPView } from "../../../src/common/SPEntities";
```
Please, use the same quotes style: either `'` or `"`

NishkalankBezawada replied:
Hello @AJIXuMuK 

Updated as suggested.

Thanks,
Nishkalank

AJIXuMuK commented on src/controls/viewPicker/ViewPicker.tsx:
```diff
@@ -0,0 +1,200 @@
+     * Set the currently selected views(s);
+     */
+    private setSelectedViews(): void {
+      this._selectedView = cloneDeep(this.props.selectedView);
```
Why do you have `selectedView` as both field and property in the state?
I don't get the point to be honest.

NishkalankBezawada replied:
This global variable is used to send the selected value to the callback function ```onSelectionChanged``` as below code did not work with state. 

```
  private onChange = (event: React.FormEvent<HTMLDivElement>, option: IDropdownOption, index?: number): void => {
    const { multiSelect, onSelectionChanged } = this.props;

    if (multiSelect) {
      let selectedViews = this.state.selectedView ? cloneDeep(this.state.selectedView) as string[] : [];

      if (option.selected) {
        selectedViews.push(option.key.toString());
      }
      else {
        selectedViews = selectedViews.filter(view => view !== option.key);
      }
      this.setState({
          selectedView : selectedViews 
      })
    }
    else {
      this.setState({
          selectedView : option.key.toString()
      })
    }
    if (onSelectionChanged) {
      onSelectionChanged(cloneDeep(this._selectedView));
    }
  }
```

Hence referred this code from ListPicker control as in below screen shot
![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/adc832dc-c6e7-4ea5-add5-0665fd042d2a)

Thanks,
Nishkalank

AJIXuMuK replied:
@NishkalankBezawada - `ListPicker` doesn't have selected lists in the state, only as a field.
Your code will work if you just pass `selectedViews` or `option.key.toString()` to the `onSelectionChanged`.
Please, leave either field or state.

NishkalankBezawada replied:
Hello @AJIXuMuK 
Your suggestion worked like a charm. Thanks for your guidance. Updated accordingly in the latest commit.

Thanks,
Nishkalank

AJIXuMuK commented on src/controls/viewPicker/ViewPicker.tsx:
```diff
@@ -0,0 +1,200 @@
+        return (
+          <div className={styles.viewPicker}>
+            {loading && <Spinner className={styles.spinner} size={SpinnerSize.xSmall}/>}
+            <Dropdown {...dropdownProps} />
```
is it shown during loading? if yes - why?

NishkalankBezawada replied:
Updated in the latest commit as suggested.

Thanks,
Nishkalank

AJIXuMuK commented on src/services/SPService.ts:
```diff
@@ -735,4 +736,55 @@ export default class SPService implements ISPService {
+  /**
+   * Gets the collection of view for a selected list
+   */
+  public async getViews(listId?: string, _webAbsoluteUrl?:string, _orderBy?: orderBy, filter?: string): Promise<ISPViews> {
```
why prefix with `_` **some** of the parameters?
Could you please use the same standard naming approach?

NishkalankBezawada replied:
Updated as suggested in the latest commit. 

Thanks,
Nishkalank

AJIXuMuK reviewed (changes requested):
@NishkalankBezawada - sorry for a long delay.
Added a few more comments.

NishkalankBezawada commented:
Hello @AJIXuMuK @joelfmrodrigues

Corrected and completed the changes dated 24-06-2023 as suggested and added two more commits to this PR. Kindly review.

Thanks,
Nishkalank

AJIXuMuK reviewed (changes requested):
@NishkalankBezawada - please modify the logic with the `_selectedView`.
And also resolve the conflicts.

NishkalankBezawada commented:
Hello @AJIXuMuK & @joelfmrodrigues 

About the merge conflicts, I need your guidance as i am not able to see those as below (note : I am targetting issue-1439-viewpicker branch)


![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/1f9d5d4a-b248-4575-8c60-5b9dd4b85ff6)

Thanks,
Nishkalank

AJIXuMuK commented:
@NishkalankBezawada you need to sync your forked repo with the origin. 
After that git pull
Then switch to your branch and git merge dev. 
You will see the conflicts

NishkalankBezawada commented:
Thanks @AJIXuMuK, seems the forked 'dev' repo is upto date. 
![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/e16cb418-2050-4e77-876c-85c7024583ae)

But, this repo isn't, and i guess i have forked it from 'master' not the 'dev'
![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/3cf3869f-14ae-4cf6-8e07-336d7a8a8010)

Please suggest.

Thanks,
Nishkalank

NishkalankBezawada commented:
Thanks alot @AJIXuMuK for your guidance, I have learnt alot of new things. 

//Nishkalank

joelfmrodrigues commented:
@NishkalankBezawada we merged a PR that makes the test control much simpler, but this means that your PR now has conflicts.
Could you please sync your branch and resolve the ControlsTest conflicts? In summary, the visibility of each control is now controlled by a switch that is associated with a state property, and the controls are only displayed if the user toggles the switch. Should be easy to update by just copying the test control you added and adding it again to the latest version of the ControlsTest.tsx file

Let me know if you run into issues

NishkalankBezawada commented:
Hello @joaojmendes 

I have now merged the conflicts on my repo. 

Thanks,
Nishkalank

NishkalankBezawada commented:
Hello @joelfmrodrigues, 

As per your suggestion, i have modified the component, and added my control to it. It works with my branch now. 

![image](https://github.com/pnp/sp-dev-fx-controls-react/assets/47456098/8b8cffac-d0dc-4cb9-a1eb-065844f6c1ac)


Thanks,
Nishkalank

joelfmrodrigues commented:
@NishkalankBezawada Sorry for the very long delay that the entire process took and for all the updates you did along it. This is now merged 🙂

NishkalankBezawada commented:
> @NishkalankBezawada Sorry for the very long delay that the entire process took and for all the updates you did along it. This is now merged 🙂

Thanks @joelfmrodrigues This was long awaited one ☺️.
